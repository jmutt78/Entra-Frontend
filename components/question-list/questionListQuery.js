import gql from 'graphql-tag';
import { perPage } from '../../config.js';

const QUESTION_LIST_QUERY = gql`
  query QUESTION_LIST_QUERY($filter: String!, $offset: Int, $limit: Int) {
    questions(
      filter: $filter
      orderBy: createdAt_DESC
      limit: $limit
      offset: $offset
    ) {
      id
      title
      description
      createdAt
      approval
      answers {
        id
        body
      }
      tags {
        id
        name
      }
      views
      upVotes
      downVotes
      askedBy {
        id
        name
        display
      }
      bookMark {
        id
      }
    }
  }
`;

export default QUESTION_LIST_QUERY;
