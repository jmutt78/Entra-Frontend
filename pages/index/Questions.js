import React from 'react';
import { Query } from 'react-apollo';
import QuestionList from '../../components/question-list';
import Error from '../../components/ErrorMessage';
import CircularProgress from '@material-ui/core/CircularProgress';
//import query from "./query";
import gql from 'graphql-tag';
var query = gql`
  query QUESTION_LIST_QUERY {
    questions(filter: "all", first: 5) {
      id
      title
      description
      createdAt

      approval
      answers {
        id
        body
      }
      tags {
        id
        name
      }
      views
      upVotes
      downVotes
      askedBy {
        id
        name
        display
      }
      bookMark {
        id
      }
    }
  }
`;

const styles = {
  container: {
    padding: '40px 20px 80px 20px',
    maxWidth: 1000,
    alignSelf: 'center',
    justifySelf: 'center',
    textAlign: 'center',
    margin: '0 auto',
    // TEMP
    overflowX: 'hidden'
  }
};

export default () => (
  <Query query={query}>
    {({ data, loading, error }) => {
      if (loading) return <CircularProgress style={{ margin: 20 }} />;
      if (error) return <Error error={error} />;
      return (
        <div style={styles.container}>
          <QuestionList
            enablePagination={false}
            questions={data.questions}
            filter={'all'}
            page="1"
            name={'Latest Questions'}
          />
        </div>
      );
    }}
  </Query>
);
