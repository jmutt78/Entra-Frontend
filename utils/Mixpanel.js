import mixpanel from 'mixpanel-browser';
mixpanel.init('bfc78222da8bc97ec71a9e3d053adab6');

let actions = {
  identify: id => {
    mixpanel.identify(id);
  },
  alias: id => {
    mixpanel.alias(id);
  },
  track: (name, props) => {
    mixpanel.track(name, props);
  },
  people: {
    set: props => {
      mixpanel.people.set(props);
    }
  }
};

export let Mixpanel = actions;
