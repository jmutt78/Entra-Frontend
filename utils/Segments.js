import React from 'react';
import Head from 'next/head';

import SEGMENT_PIXEL_1 from './segment-1.js';

export default ({ name }) => {
  return <Head>{name === 'SEGMENT_PIXEL_1' && <SEGMENT_PIXEL_1 />}</Head>;
};
