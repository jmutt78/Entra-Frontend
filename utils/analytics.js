import ReactGA from 'react-ga';
const google = 'UA-145851763-1';

export const initGA = () => {
  ReactGA.initialize(google);
};
export const logPageView = () => {
  ReactGA.set({ page: window.location.pathname });
  ReactGA.pageview(window.location.pathname);
};
export const logEvent = (category = '', action = '') => {
  if (category && action) {
    ReactGA.event({ category, action });
  }
};
export const logException = (description = '', fatal = false) => {
  if (description) {
    ReactGA.exception({ description, fatal });
  }
};
