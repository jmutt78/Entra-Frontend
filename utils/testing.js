const fakeQuestion = {
  __typename: 'Question',
  answers: [],
  approval: true,
  askedBy: [],
  bookMark: [],
  createdAt: '2019-11-08T08:20:54.504Z',
  description: '',
  downVotes: 3,
  id: '5dc525660274390007608a12',
  searchTermFoundIn: 'Question',
  tags: [],
  title: 'test1',
  upVotes: 6,
  views: 9
};

const fakeAnswer = {
  __typename: 'Answer',
  body: 'answer test 1',
  id: '5db293ee857aba00085589c9'
};

const fakeAskedby = {
  __typename: 'User',
  display: 'userTest1',
  id: '5db289a3857aba00085589b6',
  name: 'ut1'
};

const fakeBookmark = {
  __typename: 'BookMark',
  id: '5ddbdd360274390007608a2f'
};

const fakseTags = {
  __typename: 'Tag',
  id: '5db29080857aba00085589bb',
  name: 'testTag1'
};

export { fakeQuestion, fakseTags, fakeBookmark, fakeAskedby, fakeAnswer };
