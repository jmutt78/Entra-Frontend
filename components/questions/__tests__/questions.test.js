import React from 'react';
import { shallow, mount, render } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { Typography, TextField, Select, Fab, Button } from '@material-ui/core';
import { MockedProvider } from 'react-apollo/test-utils';
import 'jest-matcher-one-of';
import wait from 'waait';
import { GraphQLError } from 'graphql';
import ErrorComp from '../../ErrorMessage.js';
import Questions from '../index';
import QuestionList from '../../question-list';
import questionListQuery from '../../question-list/questionListQuery';
import { fakeQuestion, fakeAskedby } from '../../../utils/testing';

// TODO: fix this + test fetch more
// remove pagination testing
async function setup(shouldWait, shouldError = false, graphqlError = true) {
  let mocks;
  const filter = 'all';
  const props = {
    page: 1
  };

  if (!shouldError) {
    mocks = [
      {
        request: {
          query: questionListQuery,
          variables: {
            filter,
            offset: 0,
            limit: 10
          }
        },
        result: {
          data: {
            questions: [
              {
                ...fakeQuestion,
                askedBy: [{ ...fakeAskedby }]
              }
            ]
          }
        }
      }
    ];
  } else if (graphqlError) {
    mocks = [
      {
        request: {
          query: questionListQuery,
          variables: {
            filter,
            offset: 0,
            limit: 10
          }
        },
        result: {
          errors: [new GraphQLError('GraphQL Error!')]
        }
      }
    ];
  } else if (!graphqlError) {
    mocks = [
      {
        request: {
          query: questionListQuery,
          variables: {
            filter,
            offset: 0,
            limit: 10
          }
        },
        error: new Error('Network Error!')
      }
    ];
  }

  const component = mount(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Questions {...props} />
    </MockedProvider>
  );

  if (shouldWait) {
    await wait(0);

    component.update();
  }

  return {
    component: component,
    error: component.find(ErrorComp),
    questionList: component.find(QuestionList)
  };
}

describe('Questions component', () => {
  it('should render loading state initially', async () => {
    const mocks = [
      {
        request: {
          query: questionListQuery,
          variables: {
            filter: 'all',
            offset: 0,
            limit: 10
          }
        },
        result: {
          data: {
            questions: [
              {
                ...fakeQuestion,
                askedBy: [{ ...fakeAskedby }]
              }
            ]
          }
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Questions />
      </MockedProvider>
    );
    expect(
      toJSON(wrapper.find('div[MuiCircularProgress-svg]'))
    ).toMatchSnapshot();
  });

  it('should show error when graphql errors occured fetching data from server', async () => {
    const mocks = [
      {
        request: {
          query: questionListQuery,
          variables: {
            filter: 'all',
            offset: 0,
            limit: 10
          }
        },
        result: {
          errors: [new GraphQLError('GraphQL Error!')]
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Questions />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const error = wrapper.find('[data-test="graphql-error"]');
    console.log(error.text());
    expect(error.text()).toContain('Shoot!GraphQL Error!');
    expect(toJSON(error)).toMatchSnapshot();
  });

  it('should show error when network errors occured fetching data from server', async () => {
    const { error } = await setup(true, true, false);
    expect(error.at(0).text()).toMatch(/^Shoot!Network error/);
  });

  it('should render QuestionList', async () => {
    const { questionList } = await setup(true);
    expect(questionList).toHaveLength(1);
  });
});
