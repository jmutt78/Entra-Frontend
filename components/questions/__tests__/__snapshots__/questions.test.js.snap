// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions component should render loading state initially 1`] = `null`;

exports[`Questions component should show error when graphql errors occured fetching data from server 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  GraphQL Error!
</p>
`;
